### DISCUSSION TITLE
[Question][Hw6, 7] Format of input programs, expected result in hw7

### DISCUSSION THREAD

**Discussion by p51lee** [prosyslab-classroom/cs524-program-analysis#335] [Q&A]
Hello,

I have ~~two~~ three questions about hw6 and hw7:
1. Hw6
    1. In hw6, do we have to handle boolean values applied to the `sink` function, like `sink ( x < y )`? 
2. Hw7
    1. Do all functions have only one parameter?
    2. When checking example1.ll, what my implementation prints is not identical to example1.expected:
     ```console
    -|ID 11: (= (type_env \|  %w = alloca i32, align 4\|) (ptr i32))|
    -|ID 12: (= (type_env \|  %v = alloca i8, align 1\|) (ptr i8))|
    -|ID 27: (= (ptr (type_env \|  %div = sdiv i32 %2, %3\|))
    -   (type_env \|  %w = alloca i32, align 4\|))|
    -|ID 28: (= (ptr (type_env \|  %div = sdiv i32 %2, %3\|))
    -   (type_env \|  %v = alloca i8, align 1\|))|
    +|ID 13: (= (ptr i32) (type_env \|  %w = alloca i32, align 4\|))|
    +|ID 14: (= (ptr i8) (type_env \|  %v = alloca i8, align 1\|))|
    +|ID 41: (= (type_env \|  %w = alloca i32, align 4\|)
    +   (ptr (type_env \|  %div = sdiv i32 %2, %3\|)))|
    +|ID 43: (= (type_env \|  %v = alloca i8, align 1\|)
    ```
    Is it important to equalize IDs or the order of lhs and rhs for each equation between implementation result and example1.expected?

Thank you.

**Comment by goodtaeeun** [comment]
Hi, 
1. Short answer, yes. Longer answer, even if the c code may look like `sink ( x < y )`, in the LLVM IR, `x<y` is pre-caculated then a single variable is passed to the sink function. In other words, `sink ( x < y )` is no different from `sink(x)` or `sink(y)`.

2-1. Yes, in this homework, you may assume that functions only have one parameter.
2-2. No, the satisfiability is what matters. So you don't have to match the IDs and other details.

**Comment by p51lee** [comment]
Thank you for the answer.
Question 1 -> If so, I think we should also consider instructions for converting boolean(`i1`) values to integer(`i32`) in LLVM IR.

**Comment by goodtaeeun** [comment]
Hi, 

There was some misunderstanding when I replied to your question.
Programs that contain `sink(x<y)` are not within the scope of this assignment.
You may ignore any LLVM instructions that are not mentioned in Section 3 of the README.

Handling boolean values passed to `sink` requires handling LLVM instructions that should be considered nop. Thus, you don't need to support it.

Sorry for the confusion I may have caused with the previous reply.

**Comment by YoungseokCh** [comment]
Hello, I have a follow-up question for 2-1. In HW7 documentation, there is an example function type `void -> i32*`. However, since C does not support void type argument, I thought it was the type for the function without argument. Can we still assume that functions always have one parameter? @goodtaeeun

**Comment by goodtaeeun** [comment]
Hi,
I believe what you meant by `Although C does not support void-type argument`, is that using an argument within the function body that has a void type is impossible. This is true. However, we can design the analyzer to handle the functions with no arguments as if they are functions that have a single void-type argument. Example function type `void -> i32*` is regarding such case.

**Comment by jirheee** [comment]
> Hi,
> 
> 1. Short answer, yes. Longer answer, even if the c code may look like `sink ( x < y )`, in the LLVM IR, `x<y` is pre-caculated then a single variable is passed to the sink function. In other words, `sink ( x < y )` is no different from `sink(x)` or `sink(y)`.
> 
> 2-1. Yes, in this homework, you may assume that functions only have one parameter. 2-2. No, the satisfiability is what matters. So you don't have to match the IDs and other details.

So failing CI due to ID difference doesn't matter I guess..?

**Comment by goodtaeeun** [comment]
Yes, ID difference does not matter
